src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/rewriting/RewritingRuleBaseImpl.java

```diff
@@ -0,0 +1,41 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting;
+
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.impl.RuleAppliConvertTPAddIndexNLJToHashJoin;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.Stack;
+
+public abstract class RewritingRuleBaseImpl implements RewritingRule{
+    protected Set<RuleApplication> ruleApplications = new HashSet<>();
+    protected Stack<PhysicalPlan> subPlans = new Stack<>();
```

Comment:
I notice that you never clear `ruleApplications` after `determineAllPossibleApplications` has been called from the outside. Therefore, whenever `determineAllPossibleApplications` is called again from the outside, for the next plan, there may still be entries in this set. These entries should have been removed.

Ideally, `ruleApplications` and `subPlans` should not be members of this class but, instead, they are created new whenever `determineAllPossibleApplications` is called from the outside and then passed around during the recursion of an additional internal method. That way, the implementation becomes thread safe, meaning that `determineAllPossibleApplications` can even be called from multiple threads at the same time.

So, you need something like the following:

```java
    @Override
    public Set<RuleApplication> determineAllPossibleApplications( final PhysicalPlan plan ) {
        return Set<RuleApplication> _determineAllPossibleApplications( plan, new HashSet<>(), new Stack<>() );
    }

    protected Set<RuleApplication> _determineAllPossibleApplications( final PhysicalPlan plan, final Set<RuleApplication> ruleApplications, final Stack<PhysicalPlan> currentPath ) {
        // here you have the code that you currently have in determineAllPossibleApplications,
        // with the only difference that you call _determineAllPossibleApplications for the recursion
    }
```